flink-runtime/src/main/java/org/apache/flink/runtime/security/SecurityConfiguration.java

```diff
@@ -80,13 +86,21 @@ public SecurityConfiguration(Configuration flinkConf,
 			List<SecurityModuleFactory> securityModuleFactories) {
 		this.isZkSaslDisable = flinkConf.getBoolean(SecurityOptions.ZOOKEEPER_SASL_DISABLE);
 		this.keytab = flinkConf.getString(SecurityOptions.KERBEROS_LOGIN_KEYTAB);
-		this.principal = flinkConf.getString(SecurityOptions.KERBEROS_LOGIN_PRINCIPAL);
 		this.useTicketCache = flinkConf.getBoolean(SecurityOptions.KERBEROS_LOGIN_USETICKETCACHE);
 		this.loginContextNames = parseList(flinkConf.getString(SecurityOptions.KERBEROS_LOGIN_CONTEXTS));
 		this.zkServiceName = flinkConf.getString(SecurityOptions.ZOOKEEPER_SASL_SERVICE_NAME);
 		this.zkLoginContextName = flinkConf.getString(SecurityOptions.ZOOKEEPER_SASL_LOGIN_CONTEXT_NAME);
 		this.securityModuleFactories = Collections.unmodifiableList(securityModuleFactories);
 		this.flinkConfig = checkNotNull(flinkConf);
+		String rawPrincipal = flinkConf.getString(SecurityOptions.KERBEROS_LOGIN_PRINCIPAL);
+		if (rawPrincipal != null) {
+			try {
+				rawPrincipal = rawPrincipal.replace("HOSTNAME", InetAddress.getLocalHost().getCanonicalHostName());
```

Comment:
IN order to make dynamic replacement, I think we have to have some indicator on which of the 3 Kerberos principal is used:
```
user_name@REALM
host/host_name@REALM
service_name/host_name@REALM
```
and if we use `"HOSTNAME"` as reserved keyword we also need to change documentations as well